File: non_cuda_based/cpu-naive.cpp
#include <iostream>
#include <vector>
#include <map>
#include <set>

// Simplified structure for rows in the tables, holding an ID and a single integer value.
struct Row {
    int id;
    int value;
};

// Simplified binary join function that joins two tables based on matching ID and combines their values.
std::vector<Row> binary_join(const std::vector<Row>& table1, const std::vector<Row>& table2) {
    std::vector<Row> join_result;
    for (const auto& row1 : table1) {
        for (const auto& row2 : table2) {
            if (row1.id == row2.id) {
                // Combine the values from rows with matching IDs.
                Row new_row = {row1.id, row1.value + row2.value}; // Summing values for demonstration.
                join_result.push_back(new_row);
            }
        }
    }
    return join_result;
}

// Simplified k-ary join function that joins multiple tables on matching ID and combines their values.
std::vector<Row> k_ary_join(const std::vector<std::vector<Row>>& tables) {
    std::map<int, std::vector<int>> join_map;
    for (const auto& table : tables) {
        for (const auto& row : table) {
            join_map[row.id].push_back(row.value);
        }
    }

    // Create the join result by combining values for rows with the same ID across all tables.
    std::vector<Row> join_result;
    for (auto& [id, values] : join_map) {
        if (values.size() == tables.size()) { // Ensures the row is present in all tables.
            // Sum the values for demonstration.
            int combined_value = 0;
            for (int value : values) {
                combined_value += value;
            }
            join_result.push_back({id, combined_value});
        }
    }

    return join_result;
}

int main() {
    // Sample data for tables A, B, and C, now just holding simple integer values.
    std::vector<Row> table_A = {{1, 10}, {2, 20}, {3, 30}};
    std::vector<Row> table_B = {{1, 100}, {2, 200}};
    std::vector<Row> table_C = {{1, 1000}, {3, 3000}};

    // Perform binary joins A->B and then (A,B)->C.
    auto intermediate_result = binary_join(table_A, table_B);
    auto final_result_binary = binary_join(intermediate_result, table_C);

    // Perform k-ary join on A, B, and C.
    auto final_result_k_ary = k_ary_join({table_A, table_B, table_C});

    // Outputting the results.
    std::cout << "Binary join result size: " << final_result_binary.size() << std::endl;
    for (const auto& row : final_result_binary) {
        std::cout << "ID: " << row.id << ", Combined Value: " << row.value << std::endl;
    }

    std::cout << "K-ary join result size: " << final_result_k_ary.size() << std::endl;
    for (const auto& row : final_result_k_ary) {
        std::cout << "ID: " << row.id << ", Combined Value: " << row.value << std::endl;
    }

    return 0;
}

File: non_cuda_based/cpu-naive.py
from utils import *
# Let's illustrate the concept of binary join and k-ary join using a simple example with Python data structures.
# We'll simulate joining three tables (A, B, and C) with binary and k-ary approaches.

# Sample data for tables A, B, and C.
# Each table is represented as a list of dictionaries, where each dictionary represents a row in the table.
table_A = [{'id': 1, 'value_A': 'A1'}, {'id': 2, 'value_A': 'A2'}, {'id': 3, 'value_A': 'A3'}]
table_B = [{'id': 1, 'value_B': 'B1'}, {'id': 2, 'value_B': 'B2'}]
table_C = [{'id': 1, 'value_C': 'C1'}, {'id': 3, 'value_C': 'C3'}]

# Binary Join Implementation: Joining A->B and then (A,B)->C.
def binary_join(table1, table2, key):
    # Perform a simple equi-join based on the specified key.
    join_result = []
    for row1 in table1:
        for row2 in table2:
            if row1[key] == row2[key]:
                join_result.append({**row1, **row2})  # Merge dictionaries
    return join_result

# First, join A and B.
intermediate_result = binary_join(table_A, table_B, 'id')
# Then, join the intermediate result with C.
final_result_binary = binary_join(intermediate_result, table_C, 'id')

# K-ary Join Implementation: Joining A, B, and C in a single operation.
def k_ary_join(tables, key):
    # Assuming all tables have a unique key for simplicity.
    # Creating a dictionary to map keys to combined row data.
    join_map = {}
    for table in tables:
        print(f"{colors.OKGREEN} (31) {colors.END} Table: {table}")
        for row in table:
            print(f"{colors.OKCYAN} (34) {colors.END} Row: {row}")
            row_key = row[key]
            if row_key not in join_map:
                join_map[row_key] = {}
                print(f"{colors.WARNING} (38) {colors.END} Join Map: {join_map}")
            join_map[row_key].update(row)
    # Filter out incomplete joins, i.e., rows not present in all tables.
    return [row for row in join_map.values() if all(table in row for table in ['value_A', 'value_B', 'value_C'])]

# Perform k-ary join on A, B, and C.
final_result_k_ary = k_ary_join([table_A, table_B, table_C], 'id')

print(f"Result binary\n: {final_result_binary}\nResult K-Ary:\n{final_result_k_ary}")
print(f"Binary result size: {len(final_result_binary)}\nK-ary result size: {len(final_result_k_ary)}")